### ISSUE TITLE
/sdcard/osmdroid folder

### ISSUE THREAD

**Issue by 532910** [osmdroid/osmdroid#1233]
Looks like the proper place for /sdcard/osmdroid is under /sdcard/Android.
1. I don't need this folder after app purge.
2. It't a cache but not clearable by "clear cache".

**Comment by spyhunter99** [comment]
Love that it changes with every version of android

**Comment by c-e-p-x-u-o** [comment]
For example, in https://gitlab.com/axet/android-call-recorder Storage Path is ```/storage/emulated/0/Android/data/com.github.axet.audiorecorder/files``` by default so it doesn't requires storage permission, but you can change it to everything you want (/storage/emulated/0/io.mapsquare.osmcontrubutor, for example and only then you'll need to provide storage permission).

**Comment by monsieurtanuki** [comment]
I assume that an initial idea was to keep the tiles (as files or in a SQLite database) in a folder that can be reached by several apps, in order to avoid tile duplicates in distinct folders.

That being said, in an app I created recently I changed the path (`Configuration.getInstance().setOsmdroidBasePath(activity.getExternalFilesDir(null))`) in order to:
* get rid of the `WRITE_EXTERNAL_STORAGE` permission in the manifest file
* have the tiles removed when the app is uninstalled
* an additional reason would be to keep secret the scope of the tiles I need (which is not my case)

@c-e-p-x-u-o I guess using `setOsmdroidBasePath` would help get the results you want.
Then, should the default folder be changed? Rather difficult considering the app heritage.

**Comment by cbalster** [comment]
Yes, storage detection on Android is a hassle. Especially if you try to make physical sdcards available on older android versions and across manufacturers. Starting with KK things got a little more standardized and if you can live with people on older versions not being able to use their external sd, things are a little easier: [Here](https://github.com/criticalmaps/criticalmaps-android/blob/develop/app/src/main/java/de/stephanlindauer/criticalmaps/provider/StorageLocationProvider.java) you can find an example of how I'm currently handling it. `getAllWritableStorageLocations()` gets all available locations and then the one with the most free space is selected and saved in `getAndSaveBestStorageLocation()`.
For more background check these link:
https://possiblemobile.com/2014/03/android-external-storage/ 
https://commonsware.com/blog/2017/11/15/storage-situation-removable-storage.html

**Comment by alx696** [comment]
@monsieurtanuki I agree with you. Most users don't know what's in the storage, but they can see permissions we use. Rarely need to share data.

I used `setOsmdroidBasePath()` to change the dir. But `osmdroid` still in the storage! @cbalster , `setOsmdroidBasePath()` just change the lookup path?

Use `getApplicationContext().getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)` as data dir, `getApplicationContext().getExternalCacheDir()` as cache dir will better. At least let developers can modify.

A lot of temporary files been created on phone storage by app. 99% are cache file. As a SDK, why not give a good demonstration.

Use app private External Storage not need `WRITE_EXTERNAL_STORAGE` and other app still access to this files on external storage.

https://developer.android.com/guide/topics/data/data-storage#AccessingExtFiles

**Comment by cbalster** [comment]
I'm not sure I understand exactly what you're asking.
But if you want users to easily delete the cache and not have `WRITE_EXTERNAL_STORAGE`, I'd use something like the following snippet:

```
File osmdroidBasePath = new File(getCacheDir(), "osmdroid");
File osmdroidTileCache = new File(osmdroidBasePath, "tiles");
IConfigurationProvider configuration = Configuration.getInstance();
configuration.setOsmdroidBasePath(osmdroidBasePath);
configuration.setOsmdroidTileCache(osmdroidTileCache);

```

**Comment by alx696** [comment]
@cbalster Thanks for your help!

```
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    Configuration.getInstance()
        .setOsmdroidTileCache(
            getExternalCacheDir()
        );
    Configuration.getInstance()
        .setOsmdroidBasePath(
            getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)
        );

    setContentView(R.layout.activity_main);
...
```
>  No need `Configuration.getInstance().load()`. Otherwise, if execute `Configuration.getInstance().load()` before `setOsmdroidTileCache()`, empty `osmdroid/tile` will be created. Even if no `WRITE_EXTERNAL_STORAGE` permission...

https://github.com/osmdroid/osmdroid/wiki/How-to-use-the-osmdroid-library#main-activity
